### PR TITLE
ChannelPool Should create new channels in parallel.

### PR DESCRIPTION
**Pull request by sduskis**
Creating Channels in parallel is faster than creating them serially.

### CODE REVIEW

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/ChannelPool.java [hunk 1]
```diff
@@ -77,19 +85,49 @@ public ChannelPool(List<HeaderInterceptor> headerInterceptors, ChannelFactory fa
...
+   * @throws IOException if an exception occurred during Channel creation.
+   */
+  protected List<Channel> createNewChannels(int newChannelCount) throws IOException {
+    ExecutorService executorService = BigtableSessionSharedThreadPools.getInstance().getBatchThreadPool();
+    ListeningExecutorService listeningExecutor = MoreExecutors.listeningDecorator(executorService);
+
+    Callable<Channel> createChannelCallable = new Callable<Channel>() {
+      @Override
+      public Channel call() throws Exception {
+        return factory.create();
```
Comment (mgarolera) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/ChannelPool.java]: Could you add a comment with how long does it take to create a channel? In general all these optimizations should have a sense of how much time they save at the time they were implemented. They might become obsolete at some point and we will not know if we can clean them up.

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/ChannelPool.java [hunk 2]
```diff
@@ -77,19 +85,49 @@ public ChannelPool(List<HeaderInterceptor> headerInterceptors, ChannelFactory fa
...
+      }
+    };
+
+    List<ListenableFuture<Channel>> connectionFutures = new ArrayList<>(newChannelCount);
```
Comment (mgarolera) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/ChannelPool.java]: These are not connections

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/ChannelPool.java [hunk 3]
```diff
@@ -77,19 +85,49 @@ public ChannelPool(List<HeaderInterceptor> headerInterceptors, ChannelFactory fa
...
-            newChannelList.add(factory.create());
-          }
-          setChannels(newChannelList);
+          channels.set(
+              ImmutableList.<Channel> builder()
+                  .addAll(channels.get())
+                  .addAll(createNewChannels(capacity - totalSize.get()))
+                  .build()
+          );
           totalSize.set(capacity);
```
Comment (mgarolera) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/ChannelPool.java]: when is totalSize different from channels.size()?  Should ChannelPool constructor take a capacity parameter? Or at least initialize with ensureChannelCount(1)

**Comment by sduskis** [comment]
This didn't have the expected performance improvements.  Closing this.

**Pull request closed without merging**